Add support for local blob storage

ElanHasson opened the pull request:
This commit resolves #1023 

Added support for writing attachments to local (or UNC) paths.
- I have updated the work done for #1015 to support the Local storage
  option in the status page ![jabbr -
  status](https://f.cloud.github.com/assets/234704/1294568/71368ece-3086-11e3-98c7-100746bb1489.png)
- Updated the Administration page to reflect the new configuration
  values:
  -- LocalBlobStoragePath: This is the local path where files will be
  written to  (e.g: d:\files).
  -- LocalBlobStorageUriPrefix: This is the URI Prefix for the files to be
  served from (e.g: http://static.jabbr.net/files/)

Both values must be configured for IsValid() to return true on the
LocalBlobStorageHandler. Only one of the providers
(LocalBlobStorageHandler or AzureBlobStorageHandler) should be
configured at one time.

A separate site must be configured in IIS or another web server in order
to serve the static file uploads, offloading it from the JabbR server
and allowing for a much larger CDN story to follow without additional
enhancements to JabbR.

When directories get too large, administrators may change the
LocalBlobStoragePath and LocalBlobStorageUriPrefix configuration values
to change where files are written from without impacting already
uploaded files.

![jabbr - administration 1](https://f.cloud.github.com/assets/234704/1297384/f5220d1c-30e0-11e3-9e05-cae212b879be.png)
- Added several string resources in LanguageResources.resx to support
  the  Administration page additions

Here is a screenshot of it in action:

![jabbr](https://f.cloud.github.com/assets/234704/1294572/8c768d9c-3086-11e3-8806-8c01f1b7ce43.png)

ElanHasson committed:
Add support for local blob storage

Added support for writing attachments to local (or UNC) paths.

- I have updated the work done for #1015 to support the Local storage
option in the status page ![jabbr -
status](https://f.cloud.github.com/assets/234704/1294568/71368ece-3086-11e3-98c7-100746bb1489.png)

- Updated the Administration page to reflect the new configuration
values:
-- LocalBlobStoragePath: This is the local path where files will be
written to  (e.g: d:\files).
-- LocalBlobStorageUriPrefix: This is the URI Prefix for the files to be
served from (e.g: http://static.jabbr.net/files/)

Both values must be configured for IsValid() to return true on the
LocalBlobStorageHandler. Only one of the providers
(LocalBlobStorageHandler or AzureBlobStorageHandler) should be
configured at one time.

A separate site must be configured in IIS or another web server in order
to serve the static file uploads, offloading it from the JabbR server
and allowing for a much larger CDN story to follow without additional
enhancements to JabbR.

When directories get too large, administrators may change the
LocalBlobStoragePath and LocalBlobStorageUriPrefix configuration values
to change where files are written from without impacting already
uploaded files.

![jabbr - administration
page](https://f.cloud.github.com/assets/234704/1294554/f4ef9554-3085-11e3-9297-d0fca6d7ed0a.png)

- Added several string resources in LanguageResources.resx to support
the  Administration page additions

Here is a screenshot of it in action:

![jabbr](https://f.cloud.github.com/assets/234704/1294572/8c768d9c-3086-11e3-8806-8c01f1b7ce43.png)

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -340,9 +358,9 @@
     <value>FAQ</value>
   </data>
   <data name="Client_FAQMessage" xml:space="preserve">
-    <value>&lt;p&gt;Click on a user to send message.&lt;/p&gt;
```
These were changed automatically by the resource designer tool.

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -340,9 +358,9 @@
   </data>
   <data name="Client_FAQMessage" xml:space="preserve">
-    <value>&lt;p&gt;Click on a user to send message.&lt;/p&gt;
-&lt;p&gt;Type #roomname to create a link to a room.&lt;/p&gt;
```
These were changed automatically by the resource designer tool.

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -1094,7 +1112,7 @@ Use ? or type /? to display the FAQ and list of commands.</value>
     <value>The link used has expired.</value>
   </data>
   <data name="Account_ResetFailureMessage" xml:space="preserve">
-    <value>&lt;p&gt;Your password reset was unsuccessful: {0}.&lt;/p&gt;
```
These were changed automatically by the resource designer tool.

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -1107,14 +1125,14 @@ Use ? or type /? to display the FAQ and list of commands.</value>
     <value>&lt;a href="{0}"&gt;Reset your password&lt;/a&gt; if you've forgotten or lost your login information.</value>
   </data>
   <data name="Account_ResetPasswordSuccessMessage" xml:space="preserve">
-    <value>&lt;p&gt;A password reset email has been sent to the email address registered for user &lt;strong&gt;{0}&lt;/strong&gt;.&lt;/p&gt;
```
These were changed automatically by the resource designer tool.

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -1107,14 +1125,14 @@ Use ? or type /? to display the FAQ and list of commands.</value>
     <value>Request reset</value>
   </data>
   <data name="Account_ResetSuccessMessage" xml:space="preserve">
-    <value>&lt;p&gt;Your password has been reset successfully. &lt;/p&gt;
```
These were changed automatically by the resource designer tool.

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -773,7 +791,7 @@
     <value>(plus {0} hidden... click to expand)</value>
   </data>
   <data name="Chat_InitialMessages" xml:space="preserve">
-    <value>Welcome to JabbR
```
These were changed automatically by the resource designer tool.

ElanHasson commented on JabbR/LanguageResources.resx:
```diff
@@ -340,9 +358,9 @@
   <data name="Client_FAQMessage" xml:space="preserve">
-    <value>&lt;p&gt;Click on a user to send message.&lt;/p&gt;
-&lt;p&gt;Type #roomname to create a link to a room.&lt;/p&gt;
-&lt;p&gt;Use #test for testing.&lt;/p&gt;
```
These were changed automatically by the resource designer tool.

davidfowl commented:
Change the name LocalBlobXX to LocalFileSystem

davidfowl commented on JabbR/UploadHandlers/LocalBlobStorageHandler.cs:
```diff
@@ -0,0 +1,62 @@
+
+        public bool IsValid(string fileName, string contentType)
+        {
+            // Blob storage can handle any content
```
Evaluate the setttings func once

``` C#
var settings = _settingsFunc();

// Use settings here
```

davidfowl commented on JabbR/UploadHandlers/LocalBlobStorageHandler.cs:
```diff
@@ -0,0 +1,62 @@
+
+        public async Task<UploadResult> UploadFile(string fileName, string contentType, Stream stream)
+        {
+            // Randomize the filename everytime so we don't overwrite files
```
Evaluate the settings func once.

ElanHasson committed:
Rename LocalBlobStorageHandler to LocalFileSystemStorageHandler

- Rename LocalBlobStorageHandler to LocalFileSystemStorageHandler
- only evaluate settingsFunc once in LocalFileSystemStorageHandler

ElanHasson committed:
Changed settings properties

Renamed LocalBlobStoragePath and LocalBlobStorageUriPrefix to
LocalFileSystemStoragePath and LocalFileSystemStorageUriPrefix.

Updated contents and Naming of String Resources also to reflect naming
convention.

davidfowl commented on JabbR/UploadHandlers/LocalFileSystemStorageHandler.cs:
```diff
@@ -0,0 +1,66 @@
+                                "_" +
+                                Guid.NewGuid().ToString().Substring(0, 4) + Path.GetExtension(fileName);
+
+            if (!Directory.Exists(_settingsFunc().LocalFileSystemStoragePath))
```
You forgot to actually use the settings variable.

ElanHasson committed:
Updated to actually use settings instead of _settingsFunc()

Updated to actually use settings instead of _settingsFunc()

ElanHasson committed:
corrected method of randomFile generation

accidentally changed randomFile creation code.

ElanHasson commented:
Good to go.. let me know if you would like anything else corrected.
